R/teal_data-names.R

```diff
@@ -0,0 +1,41 @@
+#' Names of data sets in `teal_data` object
+#'
+#' Functions to get the names of a `teal_data` object.
+#' The names are extrapolated from the objects in the `qenv` environment and
+#' are not stored statically, unlike the normal behavior of `names()` function.
+#'
+#' Objects named with a `.` (dot) prefix will be ignored and not returned,
+#' unless `all.names` parameter is set to `TRUE`.
+#'
+#' @param x A (`teal_data`) object to access or modify.
+#'
+#' @return A character vector of names.
+#'
+#' @examples
+#' td <- teal_data(iris = iris)
+#' td <- within(td, mtcars <- mtcars)
+#' names(td)
+#'
+#' td <- within(td, .CO2 <- CO2)
+#' names(td)
+#'
+#' @export
+names.teal_data <- function(x) {
+  names_x <- names(as.environment(x))
```

Comment:
Using `names` in body which will return all objects, including with dot prefix.

I don't think it makes sense to use `ls(x, all.names = FALSE)` as that changes the way the function works.

We could add support to `ls` via some trickery, but I'm leaning against it, mainly as it will appear as masking `ls` when loading the package.

```r
#' @export
ls <- function(x, ...) {
  UseMethod("ls")
}

#' @export
ls.teal_data <- function(x, ...) {
  names_x <- base::ls(as.environment(x), ...)
  .get_sorted_names(names_x, join_keys(x), as.environment(x))
}

#' @export
ls.default <- function(x, ...) {
  base::ls(x, ...)
}
```

Reply:
I agree that `ls` would be the best. Does overriding need to come with a warning? I remember some big package was overriding `base::setequal,setdiff,intersect` and I remember we did the same with one package. It wasn't the problem with `base` directly but with other packages which were trying to do the same. Problem with `ls` is that there is `ls.str` so if you create an S3 this might be dispatched and fail because `str` is not a class. The only way is to use S4 I guess.

---

IMO overriding `names.qenv` makes sense even if not exactly the same as `names.environment`. With `qenv` we introduce a new class and methods in some cases can differ from the inherited ones. If we consider using `qenv` object in modules or it teal internals, names for me seems to be the most relevant.

Reply:
Regarding `ls.str` we either need to treat this as protected creating an exception in dispatch, or try to restrict dispatch only to current package namespace